Software

T2heke opened the pull request:
Added logic to find basket. Not sure, if orbiting works like this. Need feedback.

T2heke committed:
modified:   colors/colors.pkl
	modified:   image_processor.py
	modified:   main.py

T2heke committed:
Update main.py

T2heke committed:
Update main.py

T2heke committed:
Update main.py

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -42,47 +44,109 @@ def main_loop():
+                orbit_speed = max_motor_speed/15
+                
+                # find blue basket
+                if len(processedData.basket_b>0): # or processedData.basket_m>0)
```
You should have variable that keeps the current target basket

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -42,47 +44,109 @@ def main_loop():
+            # TODO: implement ball throwing logic
+            # drive ontop of the ball and throw it. How to know when the ball has been thrown?
+            elif state==State.THROW_BALL:
+                print(state)
```
You most likely want to change the state here to go after a new ball?

T2heke committed:
Update image_processor.py

T2heke committed:
Update main.py

T2heke commented:
Did the requested changes. So the orbiting movement is fine? Also, how to know when the ball has been thrown and should move on to finding the next ball?

T2heke committed:
Update main.py

T2heke committed:
Update main.py

Added listening to referee commands. How can I do it without a global variable?

T2heke committed:
Update main.py

T2heke committed:
modified:   colors/colors.pkl
	modified:   image_processor.py
	modified:   main.py
	modified:   motion.py

T2heke committed:
modified:   bootcamp-code/colors/colors.pkl
	modified:   bootcamp-code/main.py

T2heke committed:
modified:   colors/colors.pkl
	modified:   main.py

T2heke committed:
modified:   colors/colors.pkl
	modified:   main.py

T2heke committed:
Update main.py

commenting changes did onsite

T2heke committed:
Update main.py

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -33,56 +56,170 @@ def main_loop():
+
+            # STATE TO MOVE TO AND CENTER THE BALL
+            elif state == State.MOVE_CENTER_BALL:
+                last_state, new_state =  state_printer(state, last_state, new_state)
```
IS this method used for every state? Is it required to be in every state separately then?

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -33,56 +56,170 @@ def main_loop():
+                    continue
+                    
+                # find blue basket
+                if basketColor==basketColor.BLUE:
```
How should the basket color be swapped? Maybe you can switch the target basked based on the current target color

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -33,56 +56,170 @@ def main_loop():
+                    basket_dist_norm = (basket_max_distance-processedData.basket_b.xbasket_max_distance
+                    # proportional speed to the distance of the ball
+                    throw_motor_speed = throw_motor_speed_max/basket_dist_norm
+                    omni_motion.move(0, -1*throw_forward_speed, 0, throw_motor_speed)
```
Is it good to just drive forward? What happens if someting bumps the robot or if the robot just does not drive straight

T2heke replied:
Tried to implement the thrower speed to be proportional to distances. Should the motor speed be proportional to the basket distance?

T2heke committed:
Update main.py

T2heke committed:
modified:   main.py

T2heke committed:
modified:   main.py

T2heke committed:
modified:   main.py

T2heke committed:
modified:   main.py

T2heke committed:
Update main.py

T2heke committed:
modified:   main.py

T2heke committed:
Update main.py

T2heke committed:
modified:   main.py

T2heke committed:
Update main.py

T2heke commented:
Waiting for review.

T2heke committed:
modified:   main.py

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -33,56 +52,183 @@ def main_loop():
-                    # rotation if you want to turn
-                #print("X: ", x_speed, "Y: ", y_speed)
-                omni_motion.move(x_speed, y_speed, 0)
+                dest_x = ball_desired_x - processedData.balls[-1].x
```
I also suggest writing a method that normalizes the ball x or y locations so you have to write less duplicate code.

T2heke replied:
Added.

T2heke commented on software/bootcamp-code/main.py:
```diff
@@ -33,56 +52,182 @@ def main_loop():
+            # state to find the basket when ball is found - circle around the ball until basket is found, then move on to throwing
+            elif state==State.FIND_BASKET:
+                # if there are no balls, then find one
+                if len(processedData.balls)<1:
```
Added to the image_processor.py file "balls_exist".

T2heke commented on software/bootcamp-code/main.py:
```diff
@@ -33,56 +52,182 @@ def main_loop():
+                #x-speed (side speed) is the proportional speed of the normalised difference between ball x coordinate and basket x coordinate
+                x_speed_prop = (processedData.balls[-1].x - 0) / cam.rgb_width
+                if processedData.basket_b.exists:
+                    x_speed_prop=(processedData.balls[-1].x-processedData.basket_b.x)/cam.rgb_width
```
Added biggest_ball field to the image_processor analyze_balls method. Now it also returns the last item on the list. In the main code swapped out balls[-1] to biggest_ball.

T2heke committed:
Update image_processor.py

T2heke committed:
Update main.py

T2heke commented on software/bootcamp-code/main.py:
```diff
@@ -33,56 +52,182 @@ def main_loop():
+                if basket_dist_norm<0: continue # if the basket distance is a negative value, try again
+                else:
+		    thrower_speed_prop=basket_dist_norm*thrower_speed_range+throw_motor_speed_min
+                    omni_motion.move(-1*x_speed_prop*throw_move_speed, -1*y_speed_prop*throw_move_speed, 0, thrower_speed_prop)
```
Is this an outdated comment? Maybe I did the changes after you made the comment but forgot to publish? 
Or do you mean all of the THROW_BALL state? If so, can you be more specific?

T2heke committed:
Update main.py

T2heke committed:
Update main.py

T2heke committed:
modified:   main.py

T2heke committed:
modified:   main.py

T2heke committed:
modified:   main.py

T2heke committed:
modified:   main.py

T2heke committed:
deleted:    .main.py.swp

T2heke committed:
Update main.py

T2heke committed:
modified:   main.py

T2heke committed:
modified:   main.py

T2heke committed:
modified:   main.py

T2heke committed:
modified:   software/bootcamp-code/main.py

Akustav commented on software/bootcamp-code/image_processor.py:
```diff
@@ -20,12 +20,15 @@ def __str__(self) -> str:
-                balls=[], 
+                balls=[],
+                balls_exist=False,
+                biggest_ball=Object(exists = False),
```
Can the biggest ball not be gotten from the balls array? This variable could just be in the main loop but I guess this also works

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -3,17 +3,91 @@
+    # parse referee commands into python library (https://www.w3schools.com/python/python_json.asp)
+    referee=json.loads(command_list[-1])
+    command=referee["signal"]
+    if command=="start" and ("twelve" in referee["targets"]):
```
The name of the robot should be a separate variable. It should be easy to change when the organizers assign their own ID-s to your robot.

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -33,61 +107,196 @@ def main_loop():
+    command_list=[]
+    if referee_active:
+        # listen for referee commands if the referee is active
+        asyncio.get_event_loop().run_until_complete(listen_referee(command_list))
```
What happens if the referee server resets or there is a connection issue?

T2heke replied:
Should adding ayncio.get_event_loop().run_until_complete(listen_referee(command_list)) to the if statement fix that?

T2heke committed:
Update main.py

Logic to drive to the furthest ball if no balls are found in 15 seconds is almost done. Added also state to test_thrower. Researched how to get the referee commands working, still don't know.

T2heke committed:
Update main.py

T2heke committed:
modified:   colors/colors.pkl
	modified:   main.py

T2heke committed:
modified:   image_processor.py
	modified:   main.py

T2heke committed:
modified:   colors/colors.pkl
	modified:   main.py

T2heke committed:
Update main.py

T2heke committed:
Update main.py

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -33,61 +108,257 @@ def main_loop():
+            # State to drive to the furthest basket when in the find_ball state the ball has not been found for some time.
+            elif state==State.DRIVE_TO_OP_BASKET:
+                basket_depth = get_depth(processedData.depth_frame, 0, basket_to_drive.x)
+                if basket_depth<250: # if basket is closer than 20cm
```
I suggest more state exiting strategies, for example when it seems a ball for a few frames.

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -33,61 +108,257 @@ def main_loop():
+                print(basket_depth)
+                #print(processor.process_frame(aligned_depth=False).basket_b.distance)
+
+            elif state==State.FIND_FURTHEST_BASKET:
```
What is the point of this state? Could the drive to furthest basket not do the same since you already have a value in the basket_to_drive varible. You can just spin when not seeing the basket until you see it and then drive to it. It is quite hard not to see a basket. 

If you feel fancy you can also have a fallback case that when you dont see the basket either, it will swap baskets and if either basket is not visible, drive around and try not to drive out of the field.

T2heke replied:
Well the basket color which to drive towards changes based on this state. It finds both baskets, compares their distance and drives towards the one that is further. The state stays as long as it takes to find both baskets and then to have the furthest basket in view. Ideally the robot makes half a spin, in other case it makes a full spin.
I guess I could hardcode it and say, that the furthest basket is always the one that is further than half a court away, (distance bigger than 2000).

Akustav reviewed (commented):
Some driving state questions and distances should be calculated in the image processor.

T2heke committed:
Update image_processor.py

T2heke committed:
Update image_processor.py

T2heke committed:
Update main.py

T2heke committed:
Update main.py

T2heke committed:
modified:   image_processor.py
	modified:   main.py

T2heke committed:
modified:   main.py

T2heke committed:
Update main.py

T2heke committed:
Update main.py

Akustav commented on software/bootcamp-code/image_processor.py:
```diff
@@ -104,11 +116,28 @@ def analyze_balls(self, t_balls, fragments) -> list:
+        counter=0
+        sum=0
+        # 5x3 matrix
+        for y_m in range(y, y+4):
```
Instead of a for loop you can select a range from the 2d array, then average the values gotten from there. Something like attay[2:4,  2:4] selects a 2x2 area. Also what happens when the point is in the corners?

T2heke replied:
Used numpy mean.

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -3,17 +3,93 @@
+# method for listening to referee commands
+async def listen_referee(command_list):
+    ip_addr='192.168.3.28:8222'
+    async with websockets.connect('ws://'+ip_addr) as websocket:
```
I suggest wrapping it in a while as well in case when it has a issue, it will try to connect again

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -33,61 +109,265 @@ def main_loop():
+    # Ball desired coordinates are in the middle of the frame width and 400 from the upper edge of the frame
+    # (THE ZERO COORDINATES OF THE FRAME ARE IN THE UPPER LEFT CORNER.)
+    ball_desired_x = cam.rgb_width/2
+    ball_desired_y = 400
```
Can be comupted as a ratio to be clearer what the actual value means. cam.rgb_width*0.8 is clearer than a pixel value since it adds the context of the frame height and a percentage to the code

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -33,61 +109,265 @@ def main_loop():
+            
+            # Proccess the camera frame and get data
+            aligned=False
+            if (state==State.THROW_BALL or state==State.FIND_FURTHEST_BASKET or state==State.DRIVE_TO_OP_BASKET): aligned=True
```
Suggest using` state in (State.A, State.B, State.C)`

Akustav commented on software/bootcamp-code/main.py:
```diff
@@ -33,61 +109,265 @@ def main_loop():
-                    state=State.CENTER_BALL
+
+                # if we have a ball in view, center it
+                if processedData.balls_exist==True:
```
No need for the` == True` part

Akustav reviewed (commented):
Minor issues remain

T2heke committed:
Update .gitignore

T2heke committed:
Delete README.asciidoc

T2heke committed:
Delete software/bootcamp-code directory

T2heke committed:
Add files via upload

T2heke committed:
Create colors.pkl

T2heke committed:
Delete colors.pkl

T2heke committed:
Create c.txt

T2heke committed:
Add files via upload

T2heke committed:
Delete c.txt

T2heke committed:
Create test.py

T2heke committed:
Add files via upload

T2heke committed:
Create colors.pkl

T2heke committed:
Create README.asciidoc

T2heke committed:
Update image_processor.py

T2heke committed:
Update main.py

T2heke committed:
Update README.asciidoc

T2heke committed:
Update README.asciidoc

T2heke committed:
Delete README.md

T2heke committed:
Update README.asciidoc

T2heke committed:
Update README.asciidoc

T2heke committed:
Update README.asciidoc

T2heke committed:
Update README.asciidoc

T2heke commented:
Software documentation needs to be reviewed. Changed the file structure. Fixed some small issues.

T2heke committed:
Update motion.py

T2heke committed:
Update README.asciidoc

ReikoR commented on software/README.asciidoc:
```diff
@@ -1,2 +1,103 @@
+
+=== Running
+
+When using batteries, always check the  chargelevel of the battery. Information found https://ut-robotics.github.io/picr22-home/basketball_robot_guide/general/batteries.html?fbclid=IwAR1gUMz-IDl2nLR9_VfQ_3ubaFO5fuCyc9eYdoiWjIQNT7CG1ellCiDWb4M[here]
```
`fbclid` should not be in the URL.

T2heke replied:
removed

ReikoR commented on software/README.asciidoc:
```diff
@@ -1,2 +1,103 @@
+
+Check if the robot is connected to the same wifi as the controlling computer. Then you can disconnect the display. 
+
+Using Putty enter the name twelve as the client to connect to and enter the robot password `robot`.
```
Robot specific information should not be included in the documentation. Especially not the passwords.

T2heke committed:
Update README.asciidoc

T2heke committed:
Update README.asciidoc

Akustav commented on software/README.asciidoc:
```diff
@@ -1,2 +1,103 @@
+If you are using the referee server, set the referee variable to true (line 114) and set your team name (line 116).
+----
+    referee_active = True #true if referee server active, false if not
+    robot_id="twelve" # the name of your robot used by the referee server
```
referee IP is also important

T2heke replied:
Updated.

Akustav reviewed (changes requested):
Minor comment about the referee ip changing

T2heke committed:
Update README.asciidoc

T2heke committed:
Update README.asciidoc
